src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ActiveDirectoryAuthenticationProvider.cs

```diff
@@ -198,6 +199,49 @@ public override async Task<SqlAuthenticationToken> AcquireTokenAsync(SqlAuthenti
                 return new SqlAuthenticationToken(accessToken.Token, accessToken.ExpiresOn);
             }
 
+            if (parameters.AuthenticationMethod == SqlAuthenticationMethod.ActiveDirectoryFederatedIdentityCredentials)
```

Comment:
Heads up: This is going to conflict with #2380 and will need to be re-done to cache ClientAssertionCredential objects.

Reply:
Any ETA when will #2380 will be merged?

Reply:
> Any ETA when will #2380 be merged?

today, there was one failing test on the CI pipelines, waiting for the rerun to pass.

Reply:
@yhvicey done. It is merged.

Reply:
Looks like current implementation only supports credential types already defined in `Azure.Identity` lib - is this expected? As FIC + MSI is not an official credential type in `Azure.Identity` yet

Reply:
I'm not sure what you are asking. Why can't you do something similar to the below and cache the token credential objects being created?

Reply:
I mean, in current implementation, there's a `TokenCredentialKey` type was added as the key for caching, and one of its equality keys is `Type _tokenCredentialType`; and for all other cached token credentials, this field points to corresponding credential type in `Azure.Identity` lib, e.g. for Workload Identity:

```cs
if (parameters.AuthenticationMethod == SqlAuthenticationMethod.ActiveDirectoryWorkloadIdentity)
{
    // Cache WorkloadIdentityCredential based on authority and clientId
    TokenCredentialKey tokenCredentialKey = new(typeof(WorkloadIdentityCredential), authority, string.Empty, string.Empty, clientId);
    // If either tenant id, client id, or the token file path are not specified when fetching the token,
    // a CredentialUnavailableException will be thrown instead
    AccessToken accessToken = await GetTokenAsync(tokenCredentialKey, string.Empty, tokenRequestContext, cts.Token).ConfigureAwait(false);
    SqlClientEventSource.Log.TryTraceEvent("AcquireTokenAsync | Acquired access token for Workload Identity auth mode. Expiry Time: {0}", accessToken.ExpiresOn);
    return new SqlAuthenticationToken(accessToken.Token, accessToken.ExpiresOn);
}
```

So to follow existing code/pattern I suppose I should use a credential type representing corresponding authentication method as the cache key, right?
And if so, this PR needs to wait until we add corresponding authentication method in `Azure.Identity` lib.
So,
1. Shall we follow the same pattern (use corresponding credential type as part of the cache key) and wait until corresponding credential type have been added to `Azure.Identity`?
2. Or, use some special type reference as the cache key as a workaround?
3. Or, refactor the cache key object to support cases like this?

Reply:
In your current code, you have:
```
                    clientAssertionCredential = new ClientAssertionCredential(
                        tenantId,
                        clientId,
                        async (cancellationToken) =>
                        {
                            ManagedIdentityCredential miCredential = new ManagedIdentityCredential(managedIdentityClientId);
                            var token = await miCredential.GetTokenAsync(new TokenRequestContext(new string[] { $"api://AzureADTokenExchange/.default" })).ConfigureAwait(false);
                            return token.Token;
                        });
```

Instead of creating a new ManagedIdentityCredential, you want to use (or create) one from the cache. Then you will also cache the ClientAssertionCredential for re-use. You just need to work out what changes are needed (if any) to the TokenCredentialKey for your new logic. The _tokenCredentialType will be ClientAssertionCredential and the other values will have to be enough information to get or create the correct ManagedIdentityCredential object.

Reply:
The implementation of `WorkloadIdentityCredential` used `ClientAssertionCredential` as well so I was wondering if we should add a new credential type (just like `WorkloadIdentityCredential`). NVM - I'll use `ClientAssertionCredential` as part the cache key and finish the PR first